src/client/content_mapblock.cpp

```diff
@@ -403,12 +407,45 @@ void MapblockMeshGenerator::drawAutoLightedCuboid(aabb3f box, const f32 *txc,
 			d.Z = (j & 1) ? dz2 : dz1;
 			lights[j] = blendLight(d);
 		}
-		drawCuboid(box, tiles, tile_count, lights, txc);
+		drawCuboid(box, tiles, tile_count, lights, txc, mask);
 	} else {
-		drawCuboid(box, tiles, tile_count, nullptr, txc);
+		drawCuboid(box, tiles, tile_count, nullptr, txc, mask);
 	}
 }
 
+u8 MapblockMeshGenerator::getNodeBoxMask(const aabb3f &box, const std::vector<aabb3f> &boxes,
+	u8 solid_set, u8 sametype_set) const
+{
+	const f32 NODE_BOUNDARY = 0.5 * BS;
+
+	// for oversized nodebox, return immediately
+	if (box.MaxEdge.X > NODE_BOUNDARY ||
+			box.MinEdge.X < -NODE_BOUNDARY ||
+			box.MaxEdge.Y >  NODE_BOUNDARY ||
+			box.MinEdge.Y < -NODE_BOUNDARY ||
+			box.MaxEdge.Z >  NODE_BOUNDARY ||
+			box.MinEdge.Z < -NODE_BOUNDARY)
+		return 0;
+
+	// only skip rendering of faces at the node boundary
+	u8 solid_mask =
+			(box.MaxEdge.Y == NODE_BOUNDARY  ?  1 : 0) |
+			(box.MinEdge.Y == -NODE_BOUNDARY ?  2 : 0) |
+			(box.MaxEdge.X == NODE_BOUNDARY  ?  4 : 0) |
+			(box.MinEdge.X == -NODE_BOUNDARY ?  8 : 0) |
+			(box.MaxEdge.Z == NODE_BOUNDARY  ? 16 : 0) |
+			(box.MinEdge.Z == -NODE_BOUNDARY ? 32 : 0);
+
+	// for sametype culling, both faces must be at the node boundary
+	u8 sametype_mask =
+			((solid_mask & 3) == 3 ? 3 : 0) |
+			((solid_mask & 12) == 12 ? 12 : 0) |
+			((solid_mask & 48) == 48 ? 48 : 0);
+
+	return (solid_set & solid_mask) | (sametype_set & sametype_mask);
+}
```

Comment:
It wasn't hard to understand but I think a few comments wouldn't hurt, like this:
```
// We can (potentially) skip faces only if they touch the node boundary
u8 solid_mask = ...;

// For culling next to identical nodes, faces need to be symmetric (we don't differentiate by direction)
u8 sametype_mask = ...;

// Union with candidate nodes to get the faces we're actually skipping

Reply:
I've added comments and renamed the variables/parameters to reflect the meaning